Common/Securities/SecurityDefinitionSymbolResolver.cs

```diff
@@ -0,0 +1,198 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using QuantConnect.Configuration;
+using QuantConnect.Data.Auxiliary;
+using QuantConnect.Interfaces;
+using QuantConnect.Logging;
+using QuantConnect.Util;
+
+namespace QuantConnect.Securities
+{
+    /// <summary>
+    /// Resolves standardized security definitions such as FIGI, CUSIP, ISIN, SEDOL into
+    /// a properly mapped Lean <see cref="Symbol"/>.
+    /// </summary>
+    public class SecurityDefinitionSymbolResolver
+    {
+        private readonly Dictionary<string, MapFileResolver> _mapFileResolvers;
+        private readonly List<SecurityDefinition> _securityDefinitions;
+        private readonly IDataProvider _dataProvider;
+        
+        /// <summary>
+        /// Creates an instance of the symbol resolver
+        /// </summary>
+        /// <param name="dataProvider">Data provider used to obtain symbol mappings data</param>
+        /// <param name="securitiesDefinitionKey">Location to read the securities definition data from</param>
+        public SecurityDefinitionSymbolResolver(IDataProvider dataProvider = null, string securitiesDefinitionKey = null)
+        {
+            securitiesDefinitionKey ??= Path.Combine(Globals.DataFolder, "symbol-properties", "security-database.csv");
+            
+            _dataProvider = dataProvider ?? 
+                Composer.Instance.GetExportedValueByTypeName<IDataProvider>(
+                    Config.Get("data-provider", "QuantConnect.Lean.Engine.DataFeeds.DefaultDataProvider"));
+            
+            if (!SecurityDefinition.TryRead(_dataProvider, securitiesDefinitionKey, out _securityDefinitions))
+            {
+                Log.Error($"SecurityDefinitionSymbolResolver(): No security definitions data loaded from file: {securitiesDefinitionKey}");
+            }
+
+            _mapFileResolvers = new Dictionary<string, MapFileResolver>();
+        }
+        
+        /// <summary>
+        /// Converts CUSIP into a Lean <see cref="Symbol"/>
+        /// </summary>
+        /// <param name="cusip">
+        /// The Committee on Uniform Securities Identification Procedures (CUSIP) number of a security
+        /// </param>
+        /// <param name="tradingDate">
+        /// The date that the stock was trading at with the CUSIP provided. This is used
+        /// to get the ticker of the symbol on this date.
+        /// </param>
+        /// <returns>The Lean Symbol corresponding to the CUSIP number on the trading date provided</returns>
+        public Symbol CUSIP(string cusip, DateTime tradingDate)
+        {
+            if (string.IsNullOrWhiteSpace(cusip))
+            {
+                return null;
+            }
+            
+            return SecurityDefinitionToSymbol(
+                _securityDefinitions?.FirstOrDefault(x => x.CUSIP == cusip),
```

Comment:
Ah something I didn't notice before is that comparisons here are subject to case, they should probably ignore case to be more user friendly?

Reply:
I thought about it for a second in the middle of my previous refactor, but decided it didn't really make sense, given that everybody only uses the upper-case variants of these security identifiers.

Most likely all of these security identifiers have standardization regarding case-sensitivity. I'm not 100% sure what to do here honestly.

If we ignored case when comparing, we'd be accepting non-standard security identifier values, though it'd come at the cost of ease of use.

I'm okay with either one of these choices, just making sure that I brought it up for consideration before I make the change